src/ledger/LedgerTxnExpirationSQL.cpp

```diff
@@ -0,0 +1,382 @@
+        : mDb(db)
+    {
+        mKeyHashes.reserve(keys.size());
+        for (auto const& k : keys)
+        {
+            throwIfNotExpiration(k.type());
+            mKeyHashes.emplace_back(toOpaqueBase64(k.expiration().keyHash));
+        }
+    }
+
+    std::vector<LedgerEntry>
+    doSqliteSpecificOperation(soci::sqlite3_session_backend* sq) override
+    {
+        std::vector<char const*> cStrKeyHashes;
+        cStrKeyHashes.reserve(mKeyHashes.size());
+        for (auto const& h : mKeyHashes)
+        {
+            cStrKeyHashes.emplace_back(h.c_str());
+        }
+        std::string sql = "SELECT ledgerentry "
+                          "FROM expiration "
+                          "WHERE keyhash IN carray(?, ?, 'char*')";
+
+        auto prep = mDb.getPreparedStatement(sql);
+        auto be = prep.statement().get_backend();
+        if (be == nullptr)
+        {
+            throw std::runtime_error("no sql backend");
+        }
+        auto sqliteStatement =
+            dynamic_cast<soci::sqlite3_statement_backend*>(be);
+        auto st = sqliteStatement->stmt_;
+
+        sqlite3_reset(st);
+        sqlite3_bind_pointer(st, 1, (void*)cStrKeyHashes.data(), "carray", 0);
+        sqlite3_bind_int(st, 2, static_cast<int>(cStrKeyHashes.size()));
+        return executeAndFetch(prep.statement());
+    }
+
+#ifdef USE_POSTGRES
+    std::vector<LedgerEntry>
+    doPostgresSpecificOperation(soci::postgresql_session_backend* pg) override
+    {
+        std::string strKeyHashes;
+        marshalToPGArray(pg->conn_, strKeyHashes, mKeyHashes);
+
+        std::string sql = "WITH r AS (SELECT unnest(:v1::TEXT[])) "
+                          "SELECT ledgerentry "
+                          "FROM expiration "
+                          "WHERE (keyHash) IN (SELECT * from r)";
+
+        auto prep = mDb.getPreparedStatement(sql);
+        auto& st = prep.statement();
+        st.exchange(soci::use(strKeyHashes));
+        return executeAndFetch(st);
+    }
+#endif
+};
+
+UnorderedMap<LedgerKey, std::shared_ptr<LedgerEntry const>>
+LedgerTxnRoot::Impl::bulkLoadExpiration(
+    UnorderedSet<LedgerKey> const& keys) const
+{
+    if (!keys.empty())
+    {
+        BulkLoadExpirationOperation op(mApp.getDatabase(), keys);
+        return populateLoadedEntries(
+            keys, mApp.getDatabase().doDatabaseTypeSpecificOperation(op));
+    }
+    else
+    {
+        return {};
+    }
+}
+
+class BulkDeleteExpirationOperation : public DatabaseTypeSpecificOperation<void>
+{
+    Database& mDb;
+    LedgerTxnConsistency mCons;
+    std::vector<std::string> mKeyHashes;
+
+  public:
+    BulkDeleteExpirationOperation(Database& db, LedgerTxnConsistency cons,
+                                  std::vector<EntryIterator> const& entries)
+        : mDb(db), mCons(cons)
+    {
+        mKeyHashes.reserve(entries.size());
+        for (auto const& e : entries)
+        {
+            releaseAssert(!e.entryExists());
```

Comment:
releaseAssertOrThrow

Reply:
This is a good catch and we need to audit the SQL files, but there's an important point to note here - throwing a regular exception will lead to an abort later on. We need to throw a `NonSociRelatedException` for it to have the **_chance_** of being caught - https://github.com/stellar/stellar-core/blob/master/src/ledger/LedgerTxn.cpp#L3645-L3655.

Now using a `NonSociRelatedException` won't help in this case because it's in the bulk delete path, and [LedgerTxnRoot::Impl::commitChild](https://github.com/stellar/stellar-core/blob/bf4363684afb710b73f6c24005142bc4f27bc674/src/ledger/LedgerTxn.cpp#L2755-L2764) aborts on all exceptions. I'm wondering if we have to abort on all exceptions there or maybe we can re-throw (this is an open question and shouldn't block this PR).